plugins/tech-insights-backend/src/service/persistence/TechInsightsDatabase.ts

```diff
@@ -168,6 +185,58 @@ export class TechInsightsDatabase implements TechInsightsStore {
     return existingSchemas.find(it => it.version === sorted[0])!!;
   }
 
+  private async deleteExpiredFactsByDate(
+    tx: Transaction,
+    factRows: { id: string; entity: string }[],
+    timestamp: DateTime,
+  ) {
+    await tx<RawDbFactRow>('facts')
+      .whereIn(
+        ['id', 'entity'],
+        factRows.map(it => [it.id, it.entity]),
+      )
+      .and.where('timestamp', '<', timestamp.toISO())
+      .delete();
+  }
+
+  private async deleteExpiredFactsByNumber(
+    tx: Transaction,
+    factRows: { id: string; entity: string }[],
+    items: number,
+  ) {
+    const deletables = await tx<RawDbFactRow>('facts')
+      .whereIn(
+        ['id', 'entity'],
+        factRows.map(it => [it.id, it.entity]),
+      )
+      .and.leftJoin(
+        this.db.raw(
+          `(select *
```

Comment:
Knex is unable to handle partitioning over row numbers with custom ordering. I'm not sure which is better, mixing single statements as string and rest of the subselect as a knexified query or using the standard for the whole subquery. Maybe we can work around this detriment in tooling with the skippy query though

Reply:
Ah alright, I think it'd be best to keep the raw query as limited as possible. Helps in the future if you end up needing to write different queries depending on the DBMS.

Thoughts on going the route of using `OFFSET` tho? Imagining it might perform a bit better

Reply:
Creating an offset query is a bit tricky in this case, unless we want to make the retrieval once per entity. I'm not quite certain how we could calculate the offset for each id/entity paired row with this. Any suggestions?

Reply:
yeah true, can't see it being better than row_number tbh. On a completely separate note, is it possible to express a single delete query? :grin:

Reply:
I'll modify this to (hopefully) have on that offending line as a raw query. I'll see if we can use the first results directly in a delete clause. Left it like this for debugging purposes. The data shouldn't be a massive amount due to the nature of the lifecycle and how it affects the DB.

Reply:
I'll leave the delete DML as a separate item. For the amount of data it is handling, I don't think the complexity of additional temporary tables is justified.